### PR TITLE
Implement logic for the list of threads in Mizar GUI

### PR DESCRIPTION
**Pull request by akopich**
The list should show the list of threds we have samples for
in the descending order (that is, just like under the Sampling
tab of Orbit).

Multiple selection is allowed. When selection changes, the
state of SamplingWithFrameTrackInputWidget is updated.

For the sake of testability `SamplingWithFrameTrackInputWidget`
is templatized now. `Q_OBJECT` cannot be templatized, but its base
can, so `SamplingWithFrameTrackInputWidgetBase` was introduced.

Also, some getters are added to `BaselineAndComparison`.

`MizarPairedData` cllects some statistics upon creation now.

Tests: Compile, Unit, Manual
Bug: http://b/237532309

### CODE REVIEW

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h [hunk 1]
```diff
@@ -5,30 +5,92 @@
...
+    std::vector<std::pair<uint32_t, uint64_t>> counts_sorted(std::begin(counts), std::end(counts));
+
+    std::sort(std::begin(counts_sorted), std::end(counts_sorted),
+              [](const auto& a, const auto& b) { return a.second > b.second; });
+
+    for (const auto& [tid, unused_count] : counts_sorted) {
+      auto item = std::make_unique<QListWidgetItem>(
+          QString::fromStdString(absl::StrFormat("[%u] %s", tid, tid_to_name.at(tid))));
+      list->addItem(item.get());
+      tid_list_widget_items_to_tids_.try_emplace(std::move(item), tid);
```
Comment (antonrohr) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:86]: I am unsure if this is safe. I think with `addItem`, the `QListWidget` takes ownership of the `QListWidgetItem`. Maybe this can then lead to a situation where the item is destructed/deleted twice? I'd also like to understand this more. @beckerhe, any experience here?

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h [hunk 2]
```diff
@@ -5,30 +5,92 @@
...
+
+ private:
+  void InitTitle(const QString& name) { GetTitle()->setText(name); }
+
+  void InitThreadList(const PairedData& data) {
+    QListWidget* list = GetThreadList();
+    list->setSelectionMode(QAbstractItemView::ExtendedSelection);
+
+    absl::flat_hash_map<uint32_t, std::string> tid_to_name = data.TidToNames();
+    absl::flat_hash_map<uint32_t, uint64_t> counts = data.TidToCallstackSampleCounts();
```
Comment (antonrohr) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:75]: const refs here to avoid copies

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h [hunk 3]
```diff
@@ -5,30 +5,92 @@
...
  private:
   std::unique_ptr<Ui::SamplingWithFrameTrackInputWidget> ui_;
+
+ protected:
+  explicit SamplingWithFrameTrackInputWidgetBase(QWidget* parent);
+
+  [[nodiscard]] QLabel* GetTitle() const;
+  [[nodiscard]] QListWidget* GetThreadList() const;
+
+  absl::node_hash_map<std::unique_ptr<QListWidgetItem>, uint32_t> tid_list_widget_items_to_tids_;
```
Comment (antonrohr) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:48]: I think I would go a different way than this map here. The reason is that the lookup that you use in the cpp is suspicious (`.at(QListWidgetItem*)` seems to work, but the types are not the same).

I would subclass `QListWidgetItem` and give it a field for thread id. Then you dont need this map here, and the other problem with the ownership also goes away. You can take a look at  `OverrideMappingItem` in SymbolLocationsDialog.cpp

File: src/MizarWidgets/MizarMainWindow.cpp [hunk 4]
```diff
@@ -12,6 +12,10 @@ namespace orbit_mizar_widgets {
 
 MizarMainWindow::MizarMainWindow() : ui_(std::make_unique<Ui::MainWindow>()) { ui_->setupUi(this); }
 
+void MizarMainWindow::Init(const orbit_mizar_data::BaselineAndComparison* baseline_and_comparison) {
+  ui_->sampling_with_frame_track_widget_->Init(baseline_and_comparison);
+}
```
Comment (antonrohr) [inline src/MizarWidgets/MizarMainWindow.cpp:17]: Why pointer instead of const ref? Here and in `SamplingWithFrameTrackWidget::Init`

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h (hunk 1 above)
Comment (akopich) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:86]: > Maybe this can then lead to a situation where the item is destructed/deleted twice? 

Doesn't seem so. There was an issue with QListWidgetItem emitting signals upon destruction which lead to the need to carefully choose field destruction order. But yeah, probably, there's another way.

File: src/MizarWidgets/MizarMainWindow.cpp (hunk 4 above)
Comment (akopich) [reply src/MizarWidgets/MizarMainWindow.cpp:17]: I've spent some time contemplating the options. Pass by pointer emphasises that the ownership is not transferred at the call cite.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h (hunk 3 above)
Comment (akopich) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:48]: `.at(QListWidgetItem*)` should be fine, as `operator==` and hash is guaranteed to be the same for unique_ptr and the raw ptr. 

>I would subclass QListWidgetItem and give it a field for thread id. 

But I would have to downcast the `QListWidgetItem*` to `QTidListWidgetItem*` which IMO is more suspicious.

File: src/MizarData/MizarPairedDataTest.cpp [hunk 5]
```diff
@@ -70,8 +69,20 @@ constexpr uint64_t kRelativeTime2 = 15;
...
 constexpr uint64_t kRelativeTime4 = 30;
 constexpr uint64_t kRelativeTime5 = 40;
-constexpr uint64_t kTID = 0x3AD1;
-constexpr uint64_t kAnotherTID = 0x3AD2;
+constexpr uint64_t kRelativeTimeTooLate = 1000;
+constexpr uint32_t kTID = 0x3AD1;
+constexpr uint32_t kAnotherTID = 0x3AD2;
+constexpr uint32_t kNamelessTID = 0x3AD3;
+const std::string kThreadName = "thread";
+const std::string kOtherThreadName = "other thread";
```
Comment (beckerhe) [inline src/MizarData/MizarPairedDataTest.cpp:77]: Nit: It's good practice to wrap the literals into a std::string_view and construct the strings when needed through the explicit constructor.

File: src/MizarData/include/MizarData/BaselineAndComparison.h [hunk 6]
```diff
@@ -69,6 +69,9 @@ class BaselineAndComparisonTmpl {
         std::move(sfid_to_corrected_comparison_result));
   }
 
+  const Baseline<PairedData>& GetBaselineData() const { return baseline_; }
+  const Comparison<PairedData>& GetComparisonData() const { return comparison_; }
```
Comment (beckerhe) [inline src/MizarData/include/MizarData/BaselineAndComparison.h:73]: Nit: nodiscard

File: src/MizarData/include/MizarData/MizarPairedData.h [hunk 7]
```diff
@@ -95,6 +106,24 @@ class MizarPairedDataTmpl {
...
+        0, std::numeric_limits<uint64_t>::max(),
+        [this, &thread_names](const orbit_client_data::CallstackEvent& event) {
+          const uint32_t tid = event.thread_id();
+          tid_to_callstack_samples_counts_[tid]++;
+
+          if (tid_to_names_.contains(tid)) return;
+
+          const auto tid_to_name = thread_names.find(tid);
+          const std::string thread_name =
+              (tid_to_name != thread_names.end()) ? tid_to_name->second : "";
```
Comment (beckerhe) [inline src/MizarData/include/MizarData/MizarPairedData.h:122]: If you don't make this string const, it could be moved into the hash map.

File: src/MizarWidgets/MizarMainWindow.cpp (hunk 4 above)
Comment (beckerhe) [reply src/MizarWidgets/MizarMainWindow.cpp:17]: Would it be an option to make `baseline_and_comparison` a constructor argument? This would make it clearer that to the user of the class that the referred object needs to be kept alive until the main window gets destroyed. Having this in a function makes this a lot harder.

File: src/MizarWidgets/MizarMainWindow.cpp [hunk 8]
```diff
@@ -12,6 +12,10 @@ namespace orbit_mizar_widgets {
 
 MizarMainWindow::MizarMainWindow() : ui_(std::make_unique<Ui::MainWindow>()) { ui_->setupUi(this); }
```
Comment (beckerhe) [inline src/MizarWidgets/MizarMainWindow.cpp:13]: Sry. This is unrelated to this PR, nevertheless I wanted to mention that.

All classes that derive from QWidget should have a `QWidget* parent = nullptr` constructor parameter which is passed on to the base class.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h [hunk 9]
```diff
@@ -5,30 +5,92 @@
...
-  template <typename PairedData = orbit_mizar_data::MizarPairedData>
-  void Init(const PairedData* data, const QString& name);
+ public slots:
+  void OnThreadSelectionChanged();
 
  private:
   std::unique_ptr<Ui::SamplingWithFrameTrackInputWidget> ui_;
+
+ protected:
+  explicit SamplingWithFrameTrackInputWidgetBase(QWidget* parent);
```
Comment (beckerhe) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:43]: Usually the parent pointer gets nullptr as its default value

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h [hunk 10]
```diff
@@ -5,30 +5,92 @@
...
+
+    std::vector<std::pair<uint32_t, uint64_t>> counts_sorted(std::begin(counts), std::end(counts));
+
+    std::sort(std::begin(counts_sorted), std::end(counts_sorted),
+              [](const auto& a, const auto& b) { return a.second > b.second; });
+
+    for (const auto& [tid, unused_count] : counts_sorted) {
+      auto item = std::make_unique<QListWidgetItem>(
+          QString::fromStdString(absl::StrFormat("[%u] %s", tid, tid_to_name.at(tid))));
+      list->addItem(item.get());
```
Comment (beckerhe) [inline src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:85]: Careful. When calling addItem you are transferring ownership to the list. You can't continue to have this wrapped in a unique_ptr. because you don't really own it. `list->addItem(item.release())` would work.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h (hunk 1 above)
Comment (beckerhe) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:86]: The proper way would be to define a model.

But it also be possible to define your own role like that:

```cpp
enum CustomRoles {
  kTidRole = Qt::UserRole + 1
};
```

and then store the value as custom data in the item:

```cpp
listItem->setData(kTidRole, tid);
```

and retrieve it if needed:

```cpp
auto tid = listItem->data(kTidRole).value<uint32_t>();
```

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h (hunk 3 above)
Comment (antonrohr) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:48]: I am going to close this comment and continue in the other one

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h (hunk 1 above)
Comment (antonrohr) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:86]: In the other comment I suggested subclassing `QWidgetListItem`, because I missed this `setData` method when reading the Qt docs. I like Hennings suggestion and am in favor of implementing it this way.

Comment (antonrohr) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:86]: Also.. I still would like to understand what exactly is happening here with destruction.

File: src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h (hunk 10 above)
Comment (akopich) [reply src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h:85]: I didn't realise smth that takes an object by pointer assumes ownership. Thanks.

File: src/MizarWidgets/MizarMainWindow.cpp (hunk 4 above)
Comment (akopich) [reply src/MizarWidgets/MizarMainWindow.cpp:17]: @antonrohr it's fine for now. Soon it won't be.

**Review by antonrohr — APPROVED** [review]
LGTM!
